File: java-solution/src/utils/FastScanner.java
package utils;

import java.util.*;
import java.io.*;

public class FastScanner implements AutoCloseable {
  BufferedReader br;
  StringTokenizer st;

  public FastScanner(File f) {
    try {
      br = new BufferedReader(new FileReader(f));
    } catch (FileNotFoundException e) {
      e.printStackTrace();
    }
  }

  public String next() {
    while (st == null || !st.hasMoreTokens()) {
      try {
        st = new StringTokenizer(br.readLine());
      } catch (IOException e) {
        e.printStackTrace();
      }
    }
    return st.nextToken();
  }

  public int nextInt() {
    return Integer.parseInt(next());
  }

  @Override
  public void close() throws IOException {
    br.close();
  }
}

File: scala-solution/src/utils/FastScanner.scala
package utils

import java.io._
import java.util.StringTokenizer

case class FastScanner(f: File) {

  var br: BufferedReader = new BufferedReader(new FileReader(f))
  var st: StringTokenizer = null

  def next(): String = {
    while ( {st == null || !st.hasMoreTokens}) try
      st = new StringTokenizer(br.readLine)
    catch {
      case e: IOException =>
        e.printStackTrace()
    }
    st.nextToken
  }

  def nextInt(): Int = next().toInt
  
  def close(): Unit = {
    br.close()
  }
}
